### DISCUSSION TITLE
Tab title in location bar (without replacing url)

### DISCUSSION THREAD

**Discussion by Aris-t2** [Aris-t2/CustomJSforFx#62] [General]
I received a few requests about adding tab title to location bar in the past. 
Instead of replacing the url, the script adds the tab title into 'page action buttons' box.

![image](https://user-images.githubusercontent.com/6013569/146513758-e46d43e6-ec61-4362-8d82-af999ff4879c.png)

https://github.com/Aris-t2/CustomJSforFx/blob/master/scripts/tab_label_in_urlbar.uc.js

2021-12-21 update
- minor position adjustments (fixes normal and touch modes)
- hide item, if urlabar is open, focused or hovered

**Comment by Speravir** [comment]
> requests about adding tab title to location bar

Hmm, I am interested in having the label right beside the Panel UI button that I have moved to top left (with Custom CSS for Fx) and much free space right of it. I used to have the tab label there in the past, but it stooped working after some Firefox code changes.

Since there is no `insertAfter` I tried replacing the `inserBefore` part in [line #10](https://github.com/Aris-t2/CustomJSforFx/blob/ee5db8a5a7b6d19eb86c3b603c78f4cf5bb362d5/scripts/tab_label_in_urlbar.uc.js#L10) with `append` or `appendChild`(and the id of the Panel UI button, of course), but I wasn’t successful. :disappointed:

**Comment by Aris-t2** [comment]
You will have to run some tests, but in general it is possible to move that title item anywhere.

Replace
`document.getElementById("page-action-buttons").insertBefore(tablabel, document.getElementById("page-action-buttons").firstChild);`

with 

`document.getElementById("titlebar").insertBefore(tablabel, document.getElementById("titlebar").firstChild);`

and add this to your userChrome.css/my_userChrome.css:

```css
#tab_label_in_urlbar {
  position: absolute !important;
  display: flex !important;
  padding-inline-start: 95px !important;
  margin-bottom: -18px !important;
}
```

I also enabled `tabs_below_titlebar_above_navigation_toolbar.css` in CustomCSSforFx.

Looks in normal window mode like this:
![grafik](https://user-images.githubusercontent.com/6013569/146934711-46c3396c-1470-4638-b031-92ea659636cb.png)

**Comment by LeeBinder** [comment]
Mighty cool. I somewhat know CSS and JS a bit myself, so I can say: impressive 🥇 !

For my style in WaterFox macOS, I'm using `margin-top: 4px !important;\` for best looks. That's easy to tweak for anybody. IMPORTANT: after any change to the script, clear startup cache via 'Help -> More Troubleshooting Information'

**Comment by Speravir** [comment]
> Replace `document.getElementById("page-action-buttons").insertBefore(tablabel, document.getElementById("page-action-buttons").firstChild);`
> 
> with
> 
> `document.getElementById("titlebar").insertBefore(tablabel, document.getElementById("titlebar").firstChild);`

Ha, using `titlebar` is the trick, so simple.

Interestingly, applying this to the [first version](https://github.com/Aris-t2/CustomJSforFx/blob/cd738e445e566e81c265cc952b5620f7dce24952/scripts/tab_label_in_urlbar.uc.js) of the script does work here without any CSS.

But, there is an annoying and strange new issue with both versions: The window control buttons are not visible anymore, but do still work (the mouse over shows it, and I also tested with minimize and maximize), only the box behind them is moved to the bottom partly hiding some buttons, cf. image:

![issue](https://user-images.githubusercontent.com/899244/147015949-ff2cbf43-7225-4987-8ae4-d5166798a11f.png)

Left is the Firefox window, right is an explorer window showing, how the controls should look. Apparently, I am still on Windows 7.

I hope this is not important, but:
> I also enabled `tabs_below_titlebar_above_navigation_toolbar.css` in CustomCSSforFx.

I use  `tabs_below_navigation_toolbar.css`.

**Comment by glennpc** [comment]
I made a slight modification in the code to make the title line up better with the URL:

It was this:

#tab_label_in_urlbar {\
		  margin-block: unset !important;\
		  margin-inline: unset !important;\
		  margin-top: -2px !important;\
		}\

I changed it to this:

#tab_label_in_urlbar {\
		  margin-block: unset !important;\
		  margin-inline: unset !important;\
                  margin-top: -1px !important;\
		  margin-bottom: -2px !important;\
		}\

**Comment by Aris-t2** [comment]
@Speravir 
I can not test on Win7 atm, but I know about this glitch with Win7 + AeroGlass + LW-themes.
It is present since Firefox 4 I guess and I remember issues occurred when I was working on Noia 4 back then.

On Win7 (AeroGlass) caption buttons are hard coded and adding another item to titlebar vertical box causes the caption buttons to shift. Therefor the themes area for them shifts too. But Win7 buttons stay where the belong and the result can be seen on your screenshot. We have to solve this using position absolute.

In my tests (currently on Win 10) caption buttons do not shift. Try this code:

```js
// 'Tab label in titlebar' script for Firefox by Aris

(function() {
	
  try {

	var titlebarlabelbox = document.createXULElement("hbox");
	titlebarlabelbox.setAttribute("id","tab_label_in_titlebar_box");

	var titlebarlabel = document.createXULElement("label");
	titlebarlabel.setAttribute("id","tab_label_in_titlebar");
	titlebarlabelbox.appendChild(titlebarlabel);
	document.getElementById("titlebar").insertBefore(titlebarlabelbox, document.getElementById("titlebar").firstChild);

	updateLabel();
  
	// catch cases where tab title can change
	document.addEventListener("TabAttrModified", updateLabel, false);
	document.addEventListener('TabSelect', updateLabel, false);
	document.addEventListener('TabOpen', updateLabel, false);
	document.addEventListener('TabClose', updateLabel, false);
	document.addEventListener('load', updateLabel, false);
	document.addEventListener("DOMContentLoaded", updateLabel, false);
  
	function updateLabel() {
	  titlebarlabel.setAttribute("value",gBrowser.selectedBrowser.contentTitle);
	}
	
	var {Services} = Components.utils.import("resource://gre/modules/Services.jsm", {});
	var sss = Components.classes["@mozilla.org/content/style-sheet-service;1"].getService(Components.interfaces.nsIStyleSheetService);
	
	var uri = Services.io.newURI("data:text/css;charset=utf-8," + encodeURIComponent('\
	  \
	  #tab_label_in_titlebar_box { \
		position: absolute !important; \
		display: flex !important; \
		padding-inline-start: 95px !important; \
	  }\
	  \
	'), null, null);
  
	sss.loadAndRegisterSheet(uri, sss.AGENT_SHEET);
  
  } catch(e) {}
	
}());
```
_(works with CustomCSSforFx options `tabs_below_navigation_toolbar.css` and `tabs_below_titlebar_above_navigation_toolbar.css`)_

@glennpc 
I tested this on two Systems and get same result in compact, normal and touch modes as before. Maybe there is another combination of CSS/JS code forcing different results here?

**Comment by glennpc** [comment]
@Aris-t2
I don't think it's a conflict with any other CSS or JS scripts. I've tested it both ways with everything else disabled, still the same. What I did notice is, the modified script does not take effect until the startupCache folder is cleared out.

Here is what I am seeing.

Before the modifications:
![before](https://user-images.githubusercontent.com/54756104/147130983-1a33fb19-fbf3-49d6-a342-87600e767cc6.png)

After the modifications:
![after](https://user-images.githubusercontent.com/54756104/147131024-0d82d39e-4bc7-4792-bb26-5c74f8595ba7.png)

**Comment by Aris-t2** [comment]
I see this on your screenshot.
Does the same happen in normal and touch modes too?

**Comment by glennpc** [comment]
@Aris-t2 
Yes, looks the same with normal and touch.

**Comment by Speravir** [comment]
> I can not test on Win7 atm, but I know about this glitch with Win7 + AeroGlass + LW-themes.
> It is present since Firefox 4 I guess and I remember issues occurred when I was working on Noia 4 back then.

And now that you speak of it I suddenly remember that I once had asked you already something in regards to these buttons (blushing :blush:). Should be somewhere in the closed threads of CustomCSSforFx.

> Try this code

Works great! Thank you. I assume there are too much preconditions for styling to add it as regular script, right?

(At least I saved the url and will certainly notice updates on the *titlebar_in_urlbar* script).

> (works with CustomCSSforFx options `tabs_below_navigation_toolbar.css` and `tabs_below_titlebar_above_navigation_toolbar.css`)

And with *appbutton_in_titlebar* styles, as well, which is why you supposedly have added the start-padding.

**Comment by Aris-t2** [comment]
@Speravir 

Here is the script version for the titlebar: https://github.com/Aris-t2/CustomJSforFx/issues/55

**Comment by Speravir** [comment]
> Here is the script version for the titlebar

 :+1: :+1: I am blushing even more, especially for your idea with var settings.

**Comment by peterwx** [comment]
There is this [tab_title_url.uc.js script (by sdavidg)](https://github.com/sdavidg/firefoxChromeScripts/blob/main/scripts/dav_tab_title_url.uc.js) that displays a tooltip on the address bar.

**Reply by Speravir** [reply]
> There is this [tab_title_url.uc.js script (by sdavidg)](https://github.com/sdavidg/firefoxChromeScripts/blob/main/scripts/dav_tab_title_url.uc.js) that displays a tooltip on the address bar.

I know. I use this myself, too. @sdavidg announced his (?) scripts himself in #39. But thanks to your answer I noticed there’s an update. :smile:

Edit: I mixed up the threads. I thought your answer was in #63. Here my reaction does not make that much sense.

**Comment by LeeBinder** [comment]
Also there is Merci-chao's [Page Title in URL Bar](https://github.com/Merci-chao/userChrome.js/blob/main/PageTitle%40Merci.chao.uc.js) (thanks to xiaoxiaoflood for the [hint](https://github.com/xiaoxiaoflood/firefox-scripts/issues/105#issuecomment-1101970729)).

**Comment by LeeBinder** [comment]
@Aris-t2 updated his script on 2022-12-13:

https://github.com/Aris-t2/CustomJSforFx/blob/master/scripts/tab_label_in_urlbar.uc.js

Changes see [commit #180f4c20e3a692d38fba1291e9d1818bb1e277f6](https://github.com/Aris-t2/CustomJSforFx/commit/180f4c20e3a692d38fba1291e9d1818bb1e277f6)

~~The script stopped working in WaterFox G5, even the latest script update doesn't work. Trying right now to figure out why.~~

**Comment by kaoh08** [comment]
When the page title is very long, it expands across the entire URL bar and visually overlaps / hides the icon on the left side (e.g. the bookmark button / add-ons icon)

![Screenshot 2025-11-02 at 15 24 21](https://github.com/user-attachments/assets/af61d8e8-fb89-4332-914c-a9f9f3642dd0)

![Screenshot 2025-11-02 at 15 24 14](https://github.com/user-attachments/assets/17f80c21-7e2e-4a82-be87-6e948ad8454d)
(long titles overflow the URL bar and cover the left-side icons, hiding the add-on status and whether the page is bookmarked.)
